controller/src/main/java/io/pravega/controller/store/stream/ZKGarbageCollector.java

```diff
@@ -0,0 +1,241 @@
+                .thenCompose(x -> zkStoreHelper.getData(batchPath))
+                .thenAccept(data -> {
+                    currentBatch.set(BitConverter.readLong(data.getData(), 0));
+                })
+                .thenRun(this::initialize)
+                .whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStarted();
+                    }
+                    latch.complete(null);
+                });
+    }
+
+    @Override
+    protected void doStop() {
+        latch.thenAccept(v -> {
+            if (gcLoop.get() != null) {
+                gcLoop.get().cancel(true);
+                gcLoop.get().whenComplete((r, e) -> {
+                    if (e != null) {
+                        notifyFailed(e);
+                    } else {
+                        notifyStopped();
+                    }
+                });
+            } else {
+                notifyStopped();
+            }
+        });
+    }
+
+    public Long getLatestBatch() {
+        return currentBatch.get();
+    }
+
+    @SneakyThrows(Exception.class)
+    private void initialize() {
+        // 1. register watch on batch path
+        watch.set(registerWatch(batchPath));
+
+        // 2. attempt to acquire leadership
+        leaderLatch.set(electLeader(leaderPath, acquiredLeadership));
+
+        // 3. if this acquires leadership, then schedule periodic garbage collection.
+        // Note: This will attempt first GC only after one period elapses. This will ensure that even if there were 
+        // back to back leadership changes with each having their own clock skews, a completed transaction record
+        // will at least be present for one GC period. 
+        // The downside is, if there are controller failures before GC period, then GC keeps getting delayed perpetually. 
+        gcLoop.set(acquiredLeadership.thenCompose(x -> Futures.loop(this::isRunning, () -> Futures.delayedFuture(this::process,
+                periodInMillis, gcExecutor), gcExecutor)));
+
+        log.info("GC {} initialized", gcName);
+    }
+
+    private byte[] getBytes(long quantized) {
+        byte[] b = new byte[Long.BYTES];
+        BitConverter.writeLong(b, 0, quantized);
+        return b;
+    }
+
+    private long getLong(byte[] data) {
+        return BitConverter.readLong(data, 0);
+    }
+
+    private long computeNewBatch() {
+        return System.currentTimeMillis() / periodInMillis;
+    }
+
+    @VisibleForTesting
+    CompletableFuture<Void> process() {
+        log.info("Starting GC {}", gcName);
+        
+        // This method is called from periodic GC thread and can only be called if this is the leader.
+        // Compute new batch id.
+        long quantized = computeNewBatch();
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+        if (quantized > currentBatch.get()) {
```

Comment:
This long has been made thread safe. But it's not clear there is not a data race. I see set being called in registerWatch and doStart, neither of which call this method directly. If these can occur in parallel, you need a better synchronization mechanism. If they cannot it might be helpful to make the flow of this class more clear by having a method that chains a sequence using completableFutures so the flow is obvious.